internal/sync/cache.go

```diff
@@ -0,0 +1,371 @@
+
+type liveStateCache struct {
+	clusterCacheConfig *rest.Config
+	logger             *zap.SugaredLogger
+	// TODO: utilize `onObjectUpdated` when switch to change event triggering for sync tasking
+	// onObjectUpdated ObjectUpdatedHandler
+
+	cluster       clustercache.ClusterCache
+	cacheSettings cacheSettings
+	lock          sync.RWMutex
+}
+
+func newLiveStateCache(
+	cluster clustercache.ClusterCache,
+) LiveStateCache {
+	return &liveStateCache{
+		cluster: cluster,
+	}
+}
+
+func NewLiveStateCache(
+	clusterCacheConfig *rest.Config,
+) LiveStateCache {
+	return &liveStateCache{
+		clusterCacheConfig: clusterCacheConfig,
+	}
+}
+
+func (c *liveStateCache) getCluster() clustercache.ClusterCache {
+	c.lock.RLock()
+	if c.cluster != nil {
+		return c.cluster
+	}
+	cacheSettings := c.cacheSettings
+	c.lock.RUnlock()
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cluster != nil {
+		return c.cluster
+	}
+
+	clusterCacheConfig := c.clusterCacheConfig
+
+	// TODO: check if we need set managed namespace
+	clusterCacheOpts := []clustercache.UpdateSettingsFunc{
+		clustercache.SetListSemaphore(semaphore.NewWeighted(clusterCacheListSemaphoreSize)),
+		clustercache.SetListPageSize(clusterCacheListPageSize),
+		clustercache.SetListPageBufferSize(clusterCacheListPageBufferSize),
+		clustercache.SetWatchResyncTimeout(clusterCacheWatchResyncDuration),
+		clustercache.SetClusterSyncRetryTimeout(clusterSyncRetryTimeoutDuration),
+		clustercache.SetResyncTimeout(clusterCacheResyncDuration),
+		clustercache.SetSettings(cacheSettings.clusterSettings),
+		clustercache.SetPopulateResourceInfoHandler(c.PopulateResourceInfo),
+		clustercache.SetRetryOptions(clusterCacheAttemptLimit, clusterCacheRetryUseBackoff, isRetryableError),
+		clustercache.SetLogr(logging.NewLogrusLogger(logging.NewWithCurrentConfig())),
+	}
+
+	clusterCache := clustercache.NewClusterCache(clusterCacheConfig, clusterCacheOpts...)
+
+	// Register event handler that is executed whenever the resource gets updated in the cache
+	_ = clusterCache.OnResourceUpdated(func(newRes *clustercache.Resource, oldRes *clustercache.Resource, namespaceResources map[kube.ResourceKey]*clustercache.Resource) {
+		c.lock.RLock()
+		cacheSettings := c.cacheSettings
+		c.lock.RUnlock()
+
+		if cacheSettings.ignoreResourceUpdatesEnabled && oldRes != nil && newRes != nil && skipResourceUpdate(resInfo(oldRes), resInfo(newRes)) {
+			return
+		}
+
+		// TODO: when switch to change event triggering for sync tasking, notify
+		//       the GitSync managing the resources that there are changes happen
+	})
+
+	c.cluster = clusterCache
+
+	return clusterCache
+}
+
+func (c *liveStateCache) PopulateResourceInfo(un *unstructured.Unstructured, isRoot bool) (interface{}, bool) {
+	res := &ResourceInfo{}
+	c.lock.RLock()
+	settings := c.cacheSettings
+	c.lock.RUnlock()
+
+	res.Health, _ = health.GetResourceHealth(un, settings.clusterSettings.ResourceHealthOverride)
```

Comment:
do you know how Resource Health is being used in this cache?

Reply:
It is used to check if we need to skip a resource update when a resource change event is triggered. https://github.com/numaproj-labs/numaplane/pull/124/files#diff-f68311d50ffbcd9fd1106622bfe4fca80b3245c1b4403d2841431600cc9e7129R161